File: day02/password_philosophy.py
#!/usr/bin/env python3
from sys import argv

def letter_count(password, letter):
    cnt = 0
    for ch in password:
        if ch == letter:
            cnt += 1
    return cnt

def is_valid(password, letter, p1, p2):
    return (password[p1-1] == letter) ^ (password[p2-1] == letter)

count1 = 0
count2 = 0
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip()
        times, letter, password = line.split(' ')
        start, end = (int(n) for n in times.split('-'))
        letter = letter[0]
        lc = letter_count(password, letter)
        if start <= letter_count(password, letter) <= end:
            count1 += 1
        if is_valid(password, letter, start, end):
            count2 += 1

print('Part 1:', count1)
print('Part 2:', count2)

File: day02/password_philosophy.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.32);
use experimental qw(signatures);

sub letter_count($letter, @password_letters) {
    return scalar grep {/$letter/} @password_letters;
}

sub is_valid($letter, $p1, $p2, @password_letters) {
    return ($password_letters[$p1-1] eq $letter) ^ $password_letters[$p2-1] eq $letter;
}

my $count1 = 0;
my $count2 = 0;

while (<>) {
    chomp;
    my ($times, $letter, $password) = split;
    my ($start, $end) = split '-', $times;
    $letter =~ s/://;
    my @password_letters = split //, $password;
    $count1++ if $start <= letter_count($letter, @password_letters) <= $end;
    $count2++ if is_valid($letter, $start, $end, @password_letters);
}

say "Part 1: $count1";
say "Part 2: $count2";
